engine/verification/assigner/engine.go

```diff
@@ -0,0 +1,258 @@
+// (through the chunk assigner), and all belong to the same execution result.
+//
+// Deduplication of chunk locators is delegated to the chunks queue.
+func (e *Engine) processChunks(chunkList flow.ChunkList, resultID flow.Identifier) {
+	for _, chunk := range chunkList {
+		log := e.log.With().
+			Hex("result_id", logging.ID(resultID)).
+			Hex("chunk_id", logging.ID(chunk.ID())).
+			Uint64("chunk_index", chunk.Index).Logger()
+
+		locator := &chunks.Locator{
+			ResultID: resultID,
+			Index:    chunk.Index,
+		}
+
+		// pushes chunk locator to the chunks queue
+		ok, err := e.chunksQueue.StoreChunkLocator(locator)
+		if err != nil {
+			log.Error().Err(err).Msg("could not push chunk locator to chunks queue")
+			continue
+		}
+		if !ok {
+			log.Debug().Msg("could not push duplicate chunk locator to chunks queue")
+			continue
+		}
+
+		// notifies chunk queue consumer of a new chunk
+		e.newChunkListener.Check()
+		log.Debug().Msg("chunk locator successfully pushed to chunks queue")
+	}
+}
+
+// stakedAtBlockID checks whether this instance of verification node has staked at specified block ID.
+// It returns true and nil if verification node is staked at referenced block ID, and returns false and nil otherwise.
+// It returns false and error if it could not extract the stake of node as a verification node at the specified block.
+func (e *Engine) stakedAtBlockID(blockID flow.Identifier) (bool, error) {
+	identity, err := protocol.StakedIdentity(e.state, blockID, e.me.NodeID())
+	if err != nil {
+		return false, fmt.Errorf("could not extract staked identify of node at block %v: %w", blockID, err)
+	}
+
+	if identity.Role != flow.RoleVerification {
+		return false, fmt.Errorf("node is staked for an invalid role. expected: %s, got: %s", flow.RoleVerification, identity.Role)
+	}
+
+	staked := identity.Stake > 0
+	return staked, nil
+}
+
+// ProcessFinalizedBlock indexes the execution receipts included in the block, and handles their chunk assignments.
+// Once it is done handling all the receipts in the block, it notifies the block consumer.
+func (e *Engine) ProcessFinalizedBlock(block *flow.Block) {
+	blockID := block.ID()
+	log := e.log.With().
+		Hex("block_id", logging.ID(blockID)).
+		Int("receipt_num", len(block.Payload.Receipts)).Logger()
+
+	log.Debug().Msg("new finalized block arrived")
+
+	err := e.indexer.IndexReceipts(blockID)
+	if err != nil {
+		// TODO: consider aborting the process
+		log.Error().Err(err).Msg("could not index receipts for block")
+	}
+
+	for _, receipt := range block.Payload.Receipts {
+		e.handleExecutionReceipt(receipt, blockID)
+	}
+
+	// tells block consumer that it is done with this block
+	e.processingNotifier.Notify(blockID)
+	log.Debug().Msg("finished processing finalized block")
+}
+
+// chunkAssignments returns the list of chunks in the chunk list assigned to this verification node.
+func (e *Engine) chunkAssignments(ctx context.Context, result *flow.ExecutionResult) (flow.ChunkList, error) {
```

Comment:
this part is directly coming from `match` engine, hence, one may skip reviewing it.